{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b1daabf-c8e3-43d4-aa32-9789abcc8619",
   "metadata": {},
   "source": [
    "# Autotalker Benchmarking"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-LÃ³pez Lab\n",
    "- **Date of Creation:** 07.12.2022\n",
    "- **Date of Last Modification:** 16.12.2022"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4e5d3297-0290-47c0-846f-705e74cdb3de",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8dbd4925-7505-4d01-bf91-95b54bb8b533",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"..\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import os\n",
    "import random\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "import anndata as ad\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "\n",
    "from autotalker.models import Autotalker\n",
    "from autotalker.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                              extract_gp_dict_from_mebocost_es_interactions,\n",
    "                              extract_gp_dict_from_nichenet_ligand_target_mx,\n",
    "                              extract_gp_dict_from_omnipath_lr_interactions,\n",
    "                              filter_and_combine_gp_dict_gps)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f91f345-c297-47cb-8650-c6d9c60a8b7e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Model\n",
    "# AnnData Keys\n",
    "dataset = \"squidpy_seqfish_mouse_organogenesis\"\n",
    "counts_key = \"counts\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "adj_key = \"spatial_connectivities\"\n",
    "spatial_key = \"spatial\"\n",
    "gp_names_key = \"autotalker_gp_names\"\n",
    "active_gp_names_key = \"autotalker_active_gp_names\"\n",
    "gp_targets_mask_key = \"autotalker_gp_targets\"\n",
    "gp_sources_mask_key = \"autotalker_gp_sources\"\n",
    "latent_key = \"autotalker_latent\"\n",
    "\n",
    "## Others\n",
    "random_seed = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f6cf198-5909-4c1e-9805-5e494a63412d",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3556ca35-3edf-4c42-84b3-2fe58d5fb2f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ignore future warnings and user warnings\n",
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=UserWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define paths\n",
    "figure_path = \"../figures\"\n",
    "model_artifacts_path = \"../model_artifacts\"\n",
    "gp_data_folder_path = \"../datasets/gp_data\" # gene program data\n",
    "srt_data_folder_path = \"../datasets/srt_data\" # spatially resolved transcriptomics data\n",
    "srt_data_gold_folder_path = f\"{srt_data_folder_path}/gold\"\n",
    "nichenet_ligand_target_mx_file_path = gp_data_folder_path + \"/nichenet_ligand_target_matrix.csv\"\n",
    "omnipath_lr_interactions_file_path = gp_data_folder_path + \"/omnipath_lr_interactions.csv\"\n",
    "\n",
    "# Create required directories\n",
    "os.makedirs(figure_path, exist_ok=True)\n",
    "os.makedirs(model_artifacts_path, exist_ok=True)\n",
    "os.makedirs(\"mlruns\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc61ed2d-6ed4-45fa-8e74-b919d74000b5",
   "metadata": {},
   "source": [
    "### 1.5 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cfa64ac5-8032-4aea-b36d-4379ffd38e05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_hyperparam_benchmarking(dataset,\n",
    "                                hyperparam_option_dict,\n",
    "                                n_iters,\n",
    "                                experiment_name,\n",
    "                                gp_mask=\"combined_priors\",\n",
    "                                save_model=False):\n",
    "    # Read dataset\n",
    "    print(\"--- DATASET ---\")\n",
    "    print(f\"Using dataset {dataset}.\")\n",
    "    adata = ad.read_h5ad(f\"{srt_data_gold_folder_path}/{dataset}.h5ad\")\n",
    "    n_nodes = adata.layers[\"counts\"].shape[0]\n",
    "    n_genes = adata.layers[\"counts\"].shape[1]\n",
    "    print(f\"Number of nodes (cells): {n_nodes}.\")\n",
    "    print(f\"Number of node features (genes): {n_genes}.\")\n",
    "        \n",
    "    # Retrieve gene program mask\n",
    "    print(\"\")\n",
    "    print(\"--- GENE PROGRAMS ---\")\n",
    "    print(f\"Using '{gp_mask}' GP mask.\")\n",
    "    \n",
    "    if gp_mask == \"fc\":\n",
    "        if hyperparam_dict[\"node_label_method\"] == \"self\":\n",
    "            n_output = len(adata.var)\n",
    "            adata.varm[gp_targets_mask_key] = np.ones((hyperparam_dict[\"n_latent_fc_gps\"], n_output))\n",
    "        elif hyperparam_dict[\"node_label_method\"] != \"self\":\n",
    "            n_output = len(adata.var) * 2\n",
    "            adata.varm[gp_targets_mask_key] = np.ones((hyperparam_dict[\"n_latent_fc_gps\"], int(n_output / 2)))\n",
    "            adata.varm[gp_sources_mask_key] = np.ones((hyperparam_dict[\"n_latent_fc_gps\"], int(n_output / 2)))\n",
    "        adata.uns[gp_names_key] = np.array([f\"FC_GP_{i}\" for i in range(hyperparam_dict[\"n_latent_fc_gps\"])])\n",
    "        n_hidden_encoder = int(hyperparam_dict[\"n_latent_fc_gps\"]/2)\n",
    "    \n",
    "    elif gp_mask == \"combined_priors\":\n",
    "        nichenet_keep_target_ratio = 0.01\n",
    "        omnipath_min_curation_effort = 0\n",
    "        gp_filter_mode = \"subset\" #None\n",
    "        combine_overlap_gps = True #False\n",
    "        overlap_thresh_source_genes = 0.9\n",
    "        overlap_thresh_target_genes = 0.9\n",
    "        overlap_thresh_genes = 0.9\n",
    "        min_genes_per_gp = 1\n",
    "        min_source_genes_per_gp = 0\n",
    "        min_target_genes_per_gp = 0\n",
    "        \n",
    "        nichenet_gp_dict = extract_gp_dict_from_nichenet_ligand_target_mx(\n",
    "            keep_target_ratio=nichenet_keep_target_ratio,\n",
    "            load_from_disk=True,\n",
    "            save_to_disk=True,\n",
    "            file_path=nichenet_ligand_target_mx_file_path)\n",
    "        omnipath_gp_dict = extract_gp_dict_from_omnipath_lr_interactions(\n",
    "            min_curation_effort=omnipath_min_curation_effort,\n",
    "            load_from_disk=False,\n",
    "            save_to_disk=True,\n",
    "            file_path=omnipath_lr_interactions_file_path)\n",
    "        mebocost_gp_dict = extract_gp_dict_from_mebocost_es_interactions(\n",
    "            dir_path = \"../datasets/gp_data/metabolite_enzyme_sensor_gps/\",\n",
    "            species=\"mouse\",\n",
    "            genes_uppercase=True)\n",
    "        combined_gp_dict = dict(nichenet_gp_dict)\n",
    "        combined_gp_dict.update(omnipath_gp_dict)\n",
    "        combined_gp_dict.update(mebocost_gp_dict)\n",
    "\n",
    "        # Filter and combine gene programs\n",
    "        combined_new_gp_dict = filter_and_combine_gp_dict_gps(\n",
    "            gp_dict=combined_gp_dict,\n",
    "            gp_filter_mode=gp_filter_mode,\n",
    "            combine_overlap_gps=combine_overlap_gps,\n",
    "            overlap_thresh_source_genes=overlap_thresh_source_genes,\n",
    "            overlap_thresh_target_genes=overlap_thresh_target_genes,\n",
    "            overlap_thresh_genes=overlap_thresh_genes,\n",
    "            verbose=True)\n",
    "\n",
    "        print(f\"Number of gene programs before filtering and combining: {len(combined_gp_dict)}.\")\n",
    "        print(f\"Number of gene programs after filtering and combining: {len(combined_new_gp_dict)}.\")\n",
    "        \n",
    "        # Add the gene program dictionary as binary masks to the adata for model training\n",
    "        add_gps_from_gp_dict_to_adata(\n",
    "            gp_dict=combined_new_gp_dict,\n",
    "            adata=adata,\n",
    "            genes_uppercase=True,\n",
    "            gp_targets_mask_key=gp_targets_mask_key,\n",
    "            gp_sources_mask_key=gp_sources_mask_key,\n",
    "            gp_names_key=gp_names_key,\n",
    "            min_genes_per_gp=min_genes_per_gp,\n",
    "            min_source_genes_per_gp=min_source_genes_per_gp,\n",
    "            min_target_genes_per_gp=min_target_genes_per_gp,\n",
    "            max_genes_per_gp=None,\n",
    "            max_source_genes_per_gp=None,\n",
    "            max_target_genes_per_gp=None)\n",
    "        n_hidden_encoder = len(adata.uns[gp_names_key])\n",
    "        \n",
    "    # Summarize gene programs\n",
    "    print(f\"Number of gene programs with probed genes: {len(adata.uns['autotalker_gp_names'])}.\")\n",
    "    print(f\"Example gene programs: {random.sample(list(adata.uns['autotalker_gp_names']), 5)}.\")\n",
    "    print(f\"Number of gene program target genes: {adata.varm['autotalker_gp_targets'].sum()}.\")\n",
    "    print(f\"Number of gene program source genes: {adata.varm['autotalker_gp_sources'].sum()}.\")\n",
    "\n",
    "    # Loop `n_iters` times through combination of hyperparams \n",
    "    iters = range(n_iters)\n",
    "    hyperparams, hyperparam_values = zip(*hyperparam_option_dict.items())\n",
    "    for hyperparam_comb in itertools.product(*hyperparam_values, iters):\n",
    "        hyperparam_dict = {}\n",
    "        for hyperparam, hyperparam_value in zip(hyperparams, hyperparam_comb):\n",
    "            hyperparam_dict[hyperparam] = hyperparam_value\n",
    "        \n",
    "        experiment = mlflow.set_experiment(experiment_name)\n",
    "            \n",
    "        # Log dataset params\n",
    "        mlflow.log_param(\"dataset\", dataset)\n",
    "        mlflow.log_param(\"n_nodes\", n_nodes)\n",
    "        mlflow.log_param(\"n_genes\", n_genes)\n",
    "    \n",
    "        # Log gp mask params\n",
    "        mlflow.log_param(\"gp_mask\", gp_mask)\n",
    "        mlflow.log_param(\"nichenet_keep_target_ratio\", nichenet_keep_target_ratio)\n",
    "        mlflow.log_param(\"omnipath_min_curation_effort\", omnipath_min_curation_effort)\n",
    "        mlflow.log_param(\"gp_filter_mode\", gp_filter_mode)\n",
    "        mlflow.log_param(\"combine_overlap_gps\", combine_overlap_gps)\n",
    "        mlflow.log_param(\"overlap_thresh_source_genes\", overlap_thresh_source_genes)\n",
    "        mlflow.log_param(\"overlap_thresh_target_genes\", overlap_thresh_target_genes)\n",
    "        mlflow.log_param(\"overlap_thresh_genes\", overlap_thresh_genes)\n",
    "        mlflow.log_param(\"min_genes_per_gp\", min_genes_per_gp)\n",
    "        mlflow.log_param(\"min_source_genes_per_gp\", min_source_genes_per_gp)\n",
    "        mlflow.log_param(\"min_target_genes_per_gp\", min_target_genes_per_gp) \n",
    "        \n",
    "        # Get time for timestamping saved artifacts\n",
    "        now = datetime.now()\n",
    "        current_timestamp = now.strftime(\"%d%m%Y_%H%M%S\")\n",
    "\n",
    "        benchmark_fig_run_dir = f\"{figure_path}/{dataset}/benchmarking/{experiment_name}/{current_timestamp}\"\n",
    "        os.makedirs(benchmark_fig_run_dir, exist_ok=True)\n",
    "        benchmark_model_run_dir = f\"{model_artifacts_path}/{dataset}/benchmarking/{experiment_name}/{current_timestamp}\"\n",
    "        os.makedirs(benchmark_model_run_dir, exist_ok=True)\n",
    "        \n",
    "        # Copy adata\n",
    "        adata_run = adata.copy()\n",
    "\n",
    "        print(\"\")\n",
    "        print(\"--- SPATIAL CONNECTIVITY ---\")\n",
    "        # Compute spatial neighborhood\n",
    "        sq.gr.spatial_neighbors(adata_run,\n",
    "                                coord_type=\"generic\",\n",
    "                                spatial_key=spatial_key,\n",
    "                                radius=hyperparam_dict[\"radius\"])\n",
    "        ax = sns.histplot(np.squeeze(np.asarray(adata_run.obsp['spatial_connectivities'].sum(axis=0))),\n",
    "                          discrete=True)\n",
    "        ax.set_title(\"Node Counts Per Number of Neighbors\")\n",
    "        ax.set_ylabel(\"Node Count\")\n",
    "        ax.set_xlabel(\"Number of Neighbors\")\n",
    "        fig = ax.figure\n",
    "        fig.savefig(f\"{benchmark_fig_run_dir}/node_counts_per_n_neighs.png\",\n",
    "                    bbox_inches=\"tight\")\n",
    "        plt.close(fig)\n",
    "        avg_edges_per_node = round(adata_run.obsp['spatial_connectivities'].sum(axis=0).mean(), 2)\n",
    "        print(f\"Average number of edges per node: {avg_edges_per_node}.\")\n",
    "        n_edges = int(sp.triu(adata_run.obsp['spatial_connectivities'], k=1).sum())\n",
    "        print(f\"Number of total edges: {n_edges}.\")\n",
    "        mlflow.log_param(\"avg_n_neighbors\", avg_edges_per_node)\n",
    "        mlflow.log_param(\"n_edges\", n_edges)\n",
    "        avg_n_neighbors_rounded = int(round(avg_edges_per_node))\n",
    "\n",
    "        # Initialize model\n",
    "        print(\"\")\n",
    "        model = Autotalker(adata_run,\n",
    "                           counts_key=counts_key,\n",
    "                           adj_key=adj_key,\n",
    "                           gp_names_key=gp_names_key,\n",
    "                           active_gp_names_key=active_gp_names_key,\n",
    "                           gp_targets_mask_key=gp_targets_mask_key,\n",
    "                           gp_sources_mask_key=gp_sources_mask_key,\n",
    "                           latent_key=latent_key,\n",
    "                           include_edge_recon_loss=hyperparam_dict[\"include_edge_recon_loss\"],\n",
    "                           include_gene_expr_recon_loss=hyperparam_dict[\"include_gene_expr_recon_loss\"],\n",
    "                           gene_expr_recon_dist=hyperparam_dict[\"gene_expr_recon_dist\"],\n",
    "                           node_label_method=hyperparam_dict[\"node_label_method\"],\n",
    "                           active_gp_thresh_ratio=hyperparam_dict[\"active_gp_thresh_ratio\"],\n",
    "                           n_hidden_encoder=n_hidden_encoder,\n",
    "                           conv_layer_encoder=hyperparam_dict[\"conv_layer_encoder\"],\n",
    "                           encoder_n_attention_heads=hyperparam_dict[\"encoder_n_attention_heads\"],\n",
    "                           n_addon_gps=hyperparam_dict[\"n_addon_gps\"])\n",
    "\n",
    "        # Train model\n",
    "        print(\"\")\n",
    "        model.train(n_epochs=hyperparam_dict[\"n_epochs\"],\n",
    "                    n_epochs_all_gps=hyperparam_dict[\"n_epochs_all_gps\"],\n",
    "                    lr=hyperparam_dict[\"lr\"],\n",
    "                    lambda_edge_recon=hyperparam_dict[\"lambda_edge_recon\"],\n",
    "                    lambda_gene_expr_recon=hyperparam_dict[\"lambda_gene_expr_recon\"],\n",
    "                    lambda_group_lasso=hyperparam_dict[\"lambda_group_lasso\"],\n",
    "                    lambda_l1_addon=hyperparam_dict[\"lambda_l1_addon\"],\n",
    "                    mlflow_experiment_id=experiment.experiment_id,\n",
    "                    verbose=True)\n",
    "\n",
    "        \"\"\"\n",
    "        print(\"\")\n",
    "        # Benchmark model\n",
    "        benchmark_dict = model.run_benchmarks(\n",
    "            adata=model.adata,\n",
    "            cell_type_key=cell_type_key,\n",
    "            spatial_key=spatial_key,\n",
    "            spatial_knng_key=\"autotalker_spatial_knng\",\n",
    "            latent_knng_key=\"autotalker_latent_knng\",\n",
    "            n_neighbors=avg_n_neighbors_rounded,\n",
    "            seed=random_seed,\n",
    "            mlflow_experiment_id=experiment.experiment_id)\n",
    "        print(\"--- BENCHMARKING RESULTS ---\")\n",
    "        print(benchmark_dict)\n",
    "        \"\"\"\n",
    "        \n",
    "        # Log visualizations\n",
    "        # Use autotalker latent space for UMAP generation\n",
    "        sc.pp.neighbors(model.adata, use_rep=latent_key)\n",
    "        sc.tl.umap(model.adata, min_dist=0.3)\n",
    "        \n",
    "        # Create subplot of cell-type annotations in spatial and latent space\n",
    "        fig, axs = plt.subplots(nrows=2, ncols=1, figsize=(10, 29))\n",
    "        title = fig.suptitle(t=\"Cell-type Annotations in Spatial and Latent Space\",\n",
    "                             x=0.85,\n",
    "                             y=0.925,\n",
    "                             fontsize=20)\n",
    "        sc.pl.spatial(adata=model.adata,\n",
    "                      color=[cell_type_key],\n",
    "                      spot_size=0.03,\n",
    "                      title=\"Spatial Space\",\n",
    "                      ax=axs[0],\n",
    "                      show=False)\n",
    "        sc.pl.umap(adata=model.adata,\n",
    "                   color=[cell_type_key],\n",
    "                   title=\"Latent Space\",\n",
    "                   ax=axs[1],\n",
    "                   show=False)\n",
    "        handles, labels = axs[0].get_legend_handles_labels()\n",
    "        lgd = fig.legend(handles, labels, bbox_to_anchor=(1.225, 0.625))\n",
    "        axs[0].get_legend().remove()\n",
    "        axs[1].get_legend().remove()\n",
    "        plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "        fig.savefig(f\"{benchmark_fig_run_dir}/cell_type_annotations_in_spatial_and_latent_space.png\",\n",
    "                    bbox_extra_artists=(lgd, title),\n",
    "                    bbox_inches=\"tight\")\n",
    "        plt.close(fig)\n",
    "\n",
    "        # Create subplot of latent Leiden cluster annotations in latent and spatial space\n",
    "        resolution = 0.3\n",
    "        model.compute_latent_graph_connectivities(adata=model.adata,\n",
    "                                                  n_neighbors=int(round(avg_edges_per_node)),\n",
    "                                                  mode=\"knn\",\n",
    "                                                  seed=random_seed)\n",
    "        sc.tl.leiden(adata=model.adata,\n",
    "                     resolution=resolution,\n",
    "                     random_state=random_seed,\n",
    "                     key_added=f\"leiden_latent_{str(resolution)}\",\n",
    "                     adjacency=model.adata.obsp[\"latent_connectivities\"])\n",
    "        fig, axs = plt.subplots(nrows=2, ncols=1, figsize=(10, 20))\n",
    "        title = fig.suptitle(t=\"Latent Leiden Cluster Annotations in Latent and Spatial Space\",\n",
    "                             x=0.85,\n",
    "                             y=0.925,\n",
    "                             fontsize=20)\n",
    "        sc.pl.umap(adata=model.adata,\n",
    "                   color=[f\"leiden_latent_{str(resolution)}\"],\n",
    "                   title=f\"Latent Space\",\n",
    "                   ax=axs[0],\n",
    "                   show=False)\n",
    "        sc.pl.spatial(adata=model.adata,\n",
    "                      color=[f\"leiden_latent_{str(resolution)}\"],\n",
    "                      spot_size=0.03,\n",
    "                      title=f\"Spatial Space\",\n",
    "                      ax=axs[1],\n",
    "                      show=False)\n",
    "        handles, labels = axs[0].get_legend_handles_labels()\n",
    "        lgd = fig.legend(handles, labels, bbox_to_anchor=(1.025, 0.7))\n",
    "        axs[0].get_legend().remove()\n",
    "        axs[1].get_legend().remove()\n",
    "        plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "        fig.savefig(f\"{benchmark_fig_run_dir}/latent_leiden_cluster_annotations_in_latent_and_spatial_space.png\",\n",
    "                    bbox_extra_artists=(lgd, title),\n",
    "                    bbox_inches=\"tight\")\n",
    "        plt.close(fig)\n",
    "        \n",
    "        mlflow.log_artifacts(benchmark_fig_run_dir,\n",
    "                             current_timestamp)\n",
    "        \n",
    "        if save_model:            \n",
    "            model.save(dir_path=benchmark_model_run_dir,\n",
    "                       overwrite=True,\n",
    "                       save_adata=True,\n",
    "                       adata_file_name=f\"{dataset}.h5ad\")\n",
    "\n",
    "        mlflow.end_run()\n",
    "        print(\"--------------------\")\n",
    "        print(\"\")\n",
    "        print(\"--------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b089dc00-0251-4131-861c-308489b66790",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_benchmarking_metrics(fig_title,\n",
    "                              df,\n",
    "                              y_col_name,\n",
    "                              save_fig=False,\n",
    "                              save_dir=\"../figures\",\n",
    "                              file_name=\"benchmarking_metrics.png\"):\n",
    "    fig, axes = plt.subplots(3, 2, sharey=True, figsize=(10, 10))\n",
    "    fig.suptitle(fig_title)\n",
    "    \n",
    "    # Graph Connectivity Distance\n",
    "    sns.boxplot(data=df, ax=axes[0, 0], x=\"gcd\", y=y_col_name)\n",
    "    axes[0, 0].set_title(\"GCD\")\n",
    "\n",
    "    # Maximum Leiden Normalized Mutual Info\n",
    "    sns.boxplot(data=df, ax=axes[0, 1], x=\"mlnmi\", y=y_col_name)\n",
    "    axes[0, 1].set_title(\"MLNMI\")\n",
    "\n",
    "    # Cell-Type Affinity Distance\n",
    "    sns.boxplot(data=df, ax=axes[1, 0], x=\"cad\", y=y_col_name)\n",
    "    axes[1, 0].set_title(\"CAD\")\n",
    "\n",
    "    # Average Absolute Log Relative Cell-Type Local Inverse Simpson's Index\n",
    "    sns.boxplot(data=df, ax=axes[1, 1], x=\"arclisi\", y=y_col_name)\n",
    "    axes[1, 1].set_title(\"ARCLISI\")\n",
    "\n",
    "    # Cell Classification Accuracy\n",
    "    sns.boxplot(data=df, ax=axes[2, 0], x=\"cca\", y=y_col_name)\n",
    "    axes[2, 0].set_title(\"CCA\")\n",
    "\n",
    "    # Gene Expression Regression Mean Squared Error\n",
    "    sns.boxplot(data=df, ax=axes[2, 1], x=\"germse\", y=y_col_name)\n",
    "    axes[2, 1].set_title(\"GERMSE\")\n",
    "\n",
    "    plt.subplots_adjust(left=0.1,\n",
    "                        bottom=0.1,\n",
    "                        right=0.9,\n",
    "                        top=0.9,\n",
    "                        wspace=0.2,\n",
    "                        hspace=0.4)\n",
    "    if save_fig:\n",
    "        # Get time for timestamping saved artefacts\n",
    "        now = datetime.now()\n",
    "        current_timestamp = now.strftime(\"%d%m%Y_%H%M%S\")\n",
    "\n",
    "        os.makedirs(save_dir, exist_ok=True)\n",
    "        plt.savefig(f\"{save_dir}/{file_name}_{current_timestamp}.png\",\n",
    "                    bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7332dc1-413d-4e64-9161-38115e91c450",
   "metadata": {},
   "source": [
    "## 2. Benchmarking Experiments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8044850-0777-410b-8e1d-a4e8946ac5f7",
   "metadata": {},
   "source": [
    "### 2.1. Benchmark Number of Neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b18d6682-9985-4f2a-87b8-2623bb928b5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03094, 0.03875, 0.05102, 0.07200, 0.10241, 0.14758]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1.]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1.]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_n_neighs\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a40f9fb2-c9ff-4bf6-82f5-2f2e71335390",
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = mlflow.search_runs(experiment_names=[experiment_name],\n",
    "                          output_format=\"list\")\n",
    "\n",
    "data = []\n",
    "for run in runs:\n",
    "    data.append({**run.data.metrics, **run.data.params})\n",
    "df = pd.DataFrame.from_dict(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2de9ad4-fd1b-440d-a254-63daa1a85f62",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_benchmarking_metrics(fig_title=\"Number of Neighbors Benchmarking Metrics\",\n",
    "                          df=df,\n",
    "                          y_col_name=\"avg_n_neighbors\",\n",
    "                          save_fig=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "864a28bf-fb6f-4442-a45c-0704350dd11a",
   "metadata": {},
   "source": [
    "### 2.2. Benchmark Loss Weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "841200e2-18e7-47e4-8764-b99ca00aceb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [0, 0.2, 0.5, 1, 2, 5]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_loss_weights\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "345e4522-7bc1-4428-b1a5-dbfb0c010c19",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [0]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_loss_weights\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e7f773f-d5c1-4b4a-b6b8-6a933903916f",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [0, 0.2, 0.5, 1, 2, 5]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0.1] # [0.7]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_loss_weights\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "620a2462-ee2a-4625-9374-037400f09ed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [0, 0.2, 0.5, 1, 2, 5]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0.4] # [0.7]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_loss_weights\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3b01bdc-0b30-4dfa-a864-9a3a754a9528",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.1]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0.1, 0.4, 0.7] # [0.7]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_loss_weights\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f265e092-361b-4e53-9968-a599971d29e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.01, 0.03, 0.3]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0.1, 0.4, 0.7] # [0.7]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_loss_weights\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b1480b0-6aa2-455e-9a2f-e8c40c6619eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.01, 0.03, 0.3]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0.01] # [0.7]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_loss_weights\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fc255df-d64b-4298-8cf5-323d831be8b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.01, 0.03, 0.3]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [2]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0] # [0.7]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_loss_weights\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55e46e48-3231-447d-a7f6-eb480a0c5e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = mlflow.search_runs(experiment_names=[experiment_name],\n",
    "                          output_format=\"list\")\n",
    "\n",
    "data = []\n",
    "for run in runs:\n",
    "    data.append({**run.data.metrics, **run.data.params})\n",
    "df = pd.DataFrame.from_dict(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d640db5a-7962-41d7-a64b-dc01e5b80602",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_benchmarking_metrics(fig_title=\"Number of Neighbors Benchmarking Metrics\",\n",
    "                          df=df,\n",
    "                          y_col_name=\"avg_n_neighbors\",\n",
    "                          save_fig=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22495810-ed2c-4bbc-9e30-c03beb663d32",
   "metadata": {},
   "source": [
    "### 2.3. Benchmark Active Gene Programs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ce0f10b-0705-4b7d-b4b7-31bebf401ce0",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0., 0.01, 0.03, 0.1, 0.3]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [5]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0, 0.01, 0.03, 0.1, 0.3]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_active_gps\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73d2d658-bdfc-4ff1-a259-daf495d18943",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.05, 0.07]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [5]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_active_gps\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b84f761d-f842-46c9-9e90-7f23b2614083",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [5]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0.7]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_active_gps\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caf867b4-8e13-47b0-821d-c2be004ea7aa",
   "metadata": {},
   "source": [
    "### 2.3. Benchmark Encoder & Decoder Architecture"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea3d2308-0eb5-467b-9369-a29e2679e8a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- DATASET ---\n",
      "Using dataset squidpy_seqfish_mouse_organogenesis.\n",
      "Number of nodes (cells): 19416.\n",
      "Number of node features (genes): 351.\n",
      "\n",
      "--- GENE PROGRAMS ---\n",
      "Using 'combined_priors' GP mask.\n",
      "Removing GP 'CTF1_ligand_receptor_GP' as it is a subset of GP 'CTF1_ligand_targetgenes_GP'.\n",
      "Removing GP 'LEP_ligand_receptor_GP' as it is a subset of GP 'LEP_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL4_ligand_receptor_GP' as it is a subset of GP 'IL4_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL12A_ligand_receptor_GP' as it is a subset of GP 'IL12A_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL23A_ligand_receptor_GP' as it is a subset of GP 'IL23A_ligand_targetgenes_GP'.\n",
      "Removing GP 'CSF2_ligand_receptor_GP' as it is a subset of GP 'CSF2_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL21_ligand_receptor_GP' as it is a subset of GP 'IL21_ligand_targetgenes_GP'.\n",
      "Removing GP 'TSLP_ligand_receptor_GP' as it is a subset of GP 'TSLP_ligand_targetgenes_GP'.\n",
      "Removing GP 'GH2_ligand_receptor_GP' as it is a subset of GP 'GH2_ligand_targetgenes_GP'.\n",
      "Removing GP 'EGF_ligand_receptor_GP' as it is a subset of GP 'EGF_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL10_ligand_receptor_GP' as it is a subset of GP 'IL10_ligand_targetgenes_GP'.\n",
      "Removing GP 'TNFSF4_ligand_receptor_GP' as it is a subset of GP 'TNFSF4_ligand_targetgenes_GP'.\n",
      "Removing GP 'TNFSF18_ligand_receptor_GP' as it is a subset of GP 'TNFSF18_ligand_targetgenes_GP'.\n",
      "Removing GP 'TGFB1_ligand_receptor_GP' as it is a subset of GP 'TGFB1_ligand_targetgenes_GP'.\n",
      "Removing GP 'BMP2_ligand_receptor_GP' as it is a subset of GP 'BMP2_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL1B_ligand_receptor_GP' as it is a subset of GP 'IL1B_ligand_targetgenes_GP'.\n",
      "Removing GP 'CSH1_ligand_receptor_GP' as it is a subset of GP 'CSH1_ligand_targetgenes_GP'.\n",
      "Removing GP 'EDN1_ligand_receptor_GP' as it is a subset of GP 'EDN1_ligand_targetgenes_GP'.\n",
      "Removing GP 'EDN3_ligand_receptor_GP' as it is a subset of GP 'EDN3_ligand_targetgenes_GP'.\n",
      "Removing GP 'GCG_ligand_receptor_GP' as it is a subset of GP 'GCG_ligand_targetgenes_GP'.\n",
      "Removing GP 'HCRT_ligand_receptor_GP' as it is a subset of GP 'HCRT_ligand_targetgenes_GP'.\n",
      "Removing GP 'IGF1_ligand_receptor_GP' as it is a subset of GP 'IGF1_ligand_targetgenes_GP'.\n",
      "Removing GP 'GRP_ligand_receptor_GP' as it is a subset of GP 'GRP_ligand_targetgenes_GP'.\n",
      "Removing GP 'NTF3_ligand_receptor_GP' as it is a subset of GP 'NTF3_ligand_targetgenes_GP'.\n",
      "Removing GP 'AVP_ligand_receptor_GP' as it is a subset of GP 'AVP_ligand_targetgenes_GP'.\n",
      "Removing GP 'CSHL1_ligand_receptor_GP' as it is a subset of GP 'CSHL1_ligand_targetgenes_GP'.\n",
      "Removing GP 'PROK1_ligand_receptor_GP' as it is a subset of GP 'PROK1_ligand_targetgenes_GP'.\n",
      "Removing GP 'AGT_ligand_receptor_GP' as it is a subset of GP 'AGT_ligand_targetgenes_GP'.\n",
      "Removing GP 'GHRH_ligand_receptor_GP' as it is a subset of GP 'GHRH_ligand_targetgenes_GP'.\n",
      "Removing GP 'DLL3_ligand_receptor_GP' as it is a subset of GP 'DLL3_ligand_targetgenes_GP'.\n",
      "Removing GP 'DLL4_ligand_receptor_GP' as it is a subset of GP 'DLL4_ligand_targetgenes_GP'.\n",
      "Removing GP 'FGF23_ligand_receptor_GP' as it is a subset of GP 'FGF23_ligand_targetgenes_GP'.\n",
      "Removing GP 'GDF6_ligand_receptor_GP' as it is a subset of GP 'GDF6_ligand_targetgenes_GP'.\n",
      "Removing GP 'GDF9_ligand_receptor_GP' as it is a subset of GP 'GDF9_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL12B_ligand_receptor_GP' as it is a subset of GP 'IL12B_ligand_targetgenes_GP'.\n",
      "Removing GP 'JAG1_ligand_receptor_GP' as it is a subset of GP 'JAG1_ligand_targetgenes_GP'.\n",
      "Removing GP 'CRLF1_ligand_receptor_GP' as it is a subset of GP 'CRLF1_ligand_targetgenes_GP'.\n",
      "Removing GP 'VGF_ligand_receptor_GP' as it is a subset of GP 'VGF_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL27_ligand_receptor_GP' as it is a subset of GP 'IL27_ligand_targetgenes_GP'.\n",
      "Removing GP 'CMTM8_ligand_receptor_GP' as it is a subset of GP 'CMTM8_ligand_targetgenes_GP'.\n",
      "Removing GP 'SECTM1_ligand_receptor_GP' as it is a subset of GP 'SECTM1_ligand_targetgenes_GP'.\n",
      "Removing GP 'CMTM2_ligand_receptor_GP' as it is a subset of GP 'CMTM2_ligand_targetgenes_GP'.\n",
      "Removing GP 'BMP8A_ligand_receptor_GP' as it is a subset of GP 'BMP8A_ligand_targetgenes_GP'.\n",
      "Removing GP 'GDF3_ligand_receptor_GP' as it is a subset of GP 'GDF3_ligand_targetgenes_GP'.\n",
      "Removing GP 'EFNB3_ligand_targetgenes_GP' as it is a subset of GP 'EFNB3_ligand_receptor_GP'.\n",
      "Removing GP 'EFNB2_ligand_targetgenes_GP' as it is a subset of GP 'EFNB2_ligand_receptor_GP'.\n",
      "Removing GP 'EFNB1_ligand_targetgenes_GP' as it is a subset of GP 'EFNB1_ligand_receptor_GP'.\n",
      "Removing GP 'CCL23_ligand_targetgenes_GP' as it is a subset of GP 'CCL23_ligand_receptor_GP'.\n",
      "Removing GP 'CCL27_ligand_targetgenes_GP' as it is a subset of GP 'CCL27_ligand_receptor_GP'.\n",
      "Removing GP 'CCL28_ligand_targetgenes_GP' as it is a subset of GP 'CCL28_ligand_receptor_GP'.\n",
      "Removing GP 'CCL16_ligand_targetgenes_GP' as it is a subset of GP 'CCL16_ligand_receptor_GP'.\n",
      "Removing GP 'CXCL5_ligand_targetgenes_GP' as it is a subset of GP 'CXCL5_ligand_receptor_GP'.\n",
      "Removing GP 'PMCH_ligand_targetgenes_GP' as it is a subset of GP 'PMCH_ligand_receptor_GP'.\n",
      "Removing GP 'NMS_ligand_targetgenes_GP' as it is a subset of GP 'NMS_ligand_receptor_GP'.\n",
      "Removing GP 'NMU_ligand_targetgenes_GP' as it is a subset of GP 'NMU_ligand_receptor_GP'.\n",
      "Removing GP 'NPB_ligand_targetgenes_GP' as it is a subset of GP 'NPB_ligand_receptor_GP'.\n",
      "Removing GP 'NPW_ligand_targetgenes_GP' as it is a subset of GP 'NPW_ligand_receptor_GP'.\n",
      "Removing GP 'RLN3_ligand_targetgenes_GP' as it is a subset of GP 'RLN3_ligand_receptor_GP'.\n",
      "Removing GP 'PYY_ligand_targetgenes_GP' as it is a subset of GP 'PYY_ligand_receptor_GP'.\n",
      "Removing GP 'PPY_ligand_targetgenes_GP' as it is a subset of GP 'PPY_ligand_receptor_GP'.\n",
      "Removing GP 'NRG3_ligand_targetgenes_GP' as it is a subset of GP 'NRG3_ligand_receptor_GP'.\n",
      "Removing GP 'NRG4_ligand_targetgenes_GP' as it is a subset of GP 'NRG4_ligand_receptor_GP'.\n",
      "Removing GP 'FGF22_ligand_targetgenes_GP' as it is a subset of GP 'FGF22_ligand_receptor_GP'.\n",
      "Removing GP 'FGF13_ligand_targetgenes_GP' as it is a subset of GP 'FGF13_ligand_receptor_GP'.\n",
      "Removing GP 'FGF11_ligand_targetgenes_GP' as it is a subset of GP 'FGF11_ligand_receptor_GP'.\n",
      "Removing GP 'FGF14_ligand_targetgenes_GP' as it is a subset of GP 'FGF14_ligand_receptor_GP'.\n",
      "Removing GP 'PODXL2_ligand_targetgenes_GP' as it is a subset of GP 'PODXL2_ligand_receptor_GP'.\n",
      "Removing GP 'ADM2_ligand_targetgenes_GP' as it is a subset of GP 'ADM2_ligand_receptor_GP'.\n",
      "Removing GP 'APOB_ligand_targetgenes_GP' as it is a subset of GP 'APOB_ligand_receptor_GP'.\n",
      "Removing GP 'NPS_ligand_targetgenes_GP' as it is a subset of GP 'NPS_ligand_receptor_GP'.\n",
      "Removing GP 'PTH2_ligand_targetgenes_GP' as it is a subset of GP 'PTH2_ligand_receptor_GP'.\n",
      "Removing GP 'INSL3_ligand_targetgenes_GP' as it is a subset of GP 'INSL3_ligand_receptor_GP'.\n",
      "Removing GP 'NECTIN3_ligand_targetgenes_GP' as it is a subset of GP 'NECTIN3_ligand_receptor_GP'.\n",
      "Removing GP 'LAMA1_ligand_targetgenes_GP' as it is a subset of GP 'LAMA1_ligand_receptor_GP'.\n",
      "Removing GP 'WNT9A_ligand_targetgenes_GP' as it is a subset of GP 'WNT9A_ligand_receptor_GP'.\n",
      "Removing GP 'WNT9B_ligand_targetgenes_GP' as it is a subset of GP 'WNT9B_ligand_receptor_GP'.\n",
      "Removing GP 'WNT3_ligand_targetgenes_GP' as it is a subset of GP 'WNT3_ligand_receptor_GP'.\n",
      "Removing GP 'SOST_ligand_targetgenes_GP' as it is a subset of GP 'SOST_ligand_receptor_GP'.\n",
      "Removing GP 'EFNA1_ligand_targetgenes_GP' as it is a subset of GP 'EFNA1_ligand_receptor_GP'.\n",
      "Removing GP 'EFNA2_ligand_targetgenes_GP' as it is a subset of GP 'EFNA2_ligand_receptor_GP'.\n",
      "Removing GP 'EFNA3_ligand_targetgenes_GP' as it is a subset of GP 'EFNA3_ligand_receptor_GP'.\n",
      "Removing GP 'CCL22_ligand_targetgenes_GP' as it is a subset of GP 'CCL22_ligand_receptor_GP'.\n",
      "Removing GP 'NRG2_ligand_targetgenes_GP' as it is a subset of GP 'NRG2_ligand_receptor_GP'.\n",
      "Removing GP 'LAMC2_ligand_targetgenes_GP' as it is a subset of GP 'LAMC2_ligand_receptor_GP'.\n",
      "Removing GP 'ICAM5_ligand_targetgenes_GP' as it is a subset of GP 'ICAM5_ligand_receptor_GP'.\n",
      "Removing GP 'COMP_ligand_targetgenes_GP' as it is a subset of GP 'COMP_ligand_receptor_GP'.\n",
      "Removing GP 'CHAD_ligand_targetgenes_GP' as it is a subset of GP 'CHAD_ligand_receptor_GP'.\n",
      "Removing GP 'THBS4_ligand_targetgenes_GP' as it is a subset of GP 'THBS4_ligand_receptor_GP'.\n",
      "Removing GP 'SEMA7A_ligand_targetgenes_GP' as it is a subset of GP 'SEMA7A_ligand_receptor_GP'.\n",
      "Removing GP 'CCL8_ligand_targetgenes_GP' as it is a subset of GP 'CCL8_ligand_receptor_GP'.\n",
      "Removing GP 'CCL7_ligand_targetgenes_GP' as it is a subset of GP 'CCL7_ligand_receptor_GP'.\n",
      "Removing GP 'XCL1_ligand_targetgenes_GP' as it is a subset of GP 'XCL1_ligand_receptor_GP'.\n",
      "Removing GP 'CCL4L1_ligand_targetgenes_GP' as it is a subset of GP 'CCL4L1_ligand_receptor_GP'.\n",
      "Removing GP 'COPA_ligand_targetgenes_GP' as it is a subset of GP 'COPA_ligand_receptor_GP'.\n",
      "Removing GP 'GUCA2A_ligand_targetgenes_GP' as it is a subset of GP 'GUCA2A_ligand_receptor_GP'.\n",
      "Removing GP 'GUCA2B_ligand_targetgenes_GP' as it is a subset of GP 'GUCA2B_ligand_receptor_GP'.\n",
      "Removing GP 'NTN4_ligand_targetgenes_GP' as it is a subset of GP 'NTN4_ligand_receptor_GP'.\n",
      "Removing GP 'CD99_ligand_targetgenes_GP' as it is a subset of GP 'CD99_ligand_receptor_GP'.\n",
      "Removing GP 'SEMA3C_ligand_targetgenes_GP' as it is a subset of GP 'SEMA3C_ligand_receptor_GP'.\n",
      "Removing GP 'NECTIN2_ligand_targetgenes_GP' as it is a subset of GP 'NECTIN2_ligand_receptor_GP'.\n",
      "Removing GP 'HLA-B_ligand_targetgenes_GP' as it is a subset of GP 'HLA-B_ligand_receptor_GP'.\n",
      "Removing GP 'HLA-C_ligand_targetgenes_GP' as it is a subset of GP 'HLA-C_ligand_receptor_GP'.\n",
      "Removing GP 'CNTN1_ligand_targetgenes_GP' as it is a subset of GP 'CNTN1_ligand_receptor_GP'.\n",
      "Removing GP 'SEMA3F_ligand_targetgenes_GP' as it is a subset of GP 'SEMA3F_ligand_receptor_GP'.\n",
      "Removing GP 'FAM3C_ligand_targetgenes_GP' as it is a subset of GP 'FAM3C_ligand_receptor_GP'.\n",
      "Removing GP 'ULBP2_ligand_targetgenes_GP' as it is a subset of GP 'ULBP2_ligand_receptor_GP'.\n",
      "Removing GP 'GPHA2_ligand_targetgenes_GP' as it is a subset of GP 'GPHA2_ligand_receptor_GP'.\n",
      "Removing GP 'FLRT3_ligand_targetgenes_GP' as it is a subset of GP 'FLRT3_ligand_receptor_GP'.\n",
      "Removing GP 'PRND_ligand_targetgenes_GP' as it is a subset of GP 'PRND_ligand_receptor_GP'.\n",
      "Removing GP 'NDP_ligand_targetgenes_GP' as it is a subset of GP 'NDP_ligand_receptor_GP'.\n",
      "Removing GP 'RSPO2_ligand_targetgenes_GP' as it is a subset of GP 'RSPO2_ligand_receptor_GP'.\n",
      "Removing GP 'RSPO4_ligand_targetgenes_GP' as it is a subset of GP 'RSPO4_ligand_receptor_GP'.\n",
      "Removing GP 'NTN3_ligand_targetgenes_GP' as it is a subset of GP 'NTN3_ligand_receptor_GP'.\n",
      "Removing GP 'DHH_ligand_targetgenes_GP' as it is a subset of GP 'DHH_ligand_receptor_GP'.\n",
      "Removing GP 'DSC1_ligand_targetgenes_GP' as it is a subset of GP 'DSC1_ligand_receptor_GP'.\n",
      "Removing GP 'DSC2_ligand_targetgenes_GP' as it is a subset of GP 'DSC2_ligand_receptor_GP'.\n",
      "Removing GP 'NLGN2_ligand_targetgenes_GP' as it is a subset of GP 'NLGN2_ligand_receptor_GP'.\n",
      "Removing GP 'NXPH1_ligand_targetgenes_GP' as it is a subset of GP 'NXPH1_ligand_receptor_GP'.\n",
      "Removing GP 'NPFF_ligand_targetgenes_GP' as it is a subset of GP 'NPFF_ligand_receptor_GP'.\n",
      "Removing GP 'TAC3_ligand_targetgenes_GP' as it is a subset of GP 'TAC3_ligand_receptor_GP'.\n",
      "Removing GP 'TAC4_ligand_targetgenes_GP' as it is a subset of GP 'TAC4_ligand_receptor_GP'.\n",
      "Removing GP 'NMB_ligand_targetgenes_GP' as it is a subset of GP 'NMB_ligand_receptor_GP'.\n",
      "Removing GP 'GALP_ligand_targetgenes_GP' as it is a subset of GP 'GALP_ligand_receptor_GP'.\n",
      "Removing GP 'MLN_ligand_targetgenes_GP' as it is a subset of GP 'MLN_ligand_receptor_GP'.\n",
      "Removing GP 'IL17B_ligand_targetgenes_GP' as it is a subset of GP 'IL17B_ligand_receptor_GP'.\n",
      "Removing GP 'QRFP_ligand_targetgenes_GP' as it is a subset of GP 'QRFP_ligand_receptor_GP'.\n",
      "Removing GP 'TXLNA_ligand_targetgenes_GP' as it is a subset of GP 'TXLNA_ligand_receptor_GP'.\n",
      "Removing GP 'Deoxyinosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Adenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Deoxyadenosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Adenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Guanosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Adenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Inosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Adenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Corticosterone_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Cortisol_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Deoxyinosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Deoxyadenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'D-Galactose_metabolite_enzyme_sensor_GP' as it is a subset of GP 'D-Glucose_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'D-Maltose_metabolite_enzyme_sensor_GP' as it is a subset of GP 'D-Glucose_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Estriol_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Estrone_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Estriol_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Estradiol_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'L-Dopa_metabolite_enzyme_sensor_GP' as it is a subset of GP 'L-Tyrosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Deoxyinosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Inosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Deoxyadenosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Inosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Guanosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Inosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Stearic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Oleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Oleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Oleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Epinephrine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Norepinephrine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Stearic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Palmitic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Palmitic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Palmitic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Androsterone_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Testosterone_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Liothyronine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Thyroxine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Cytidine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Uridine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Thymine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Uracil_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'alpha-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Linoleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Linoleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Linoleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Stearic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Stearic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'alpha-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Arachidonic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Arachidonic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Arachidonic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Prostaglandin F2a_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Prostaglandin E2_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP '13-cis-Retinoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Retinal_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'alpha-Linolenic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'alpha-Linolenic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP '13-cis-Retinoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'all-trans-Retinoic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Docosahexaenoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Eicosapentaenoic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Eicosapentaenoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Docosahexaenoic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP '13-cis-Retinoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP '9-cis-Retinoic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Calcitriol_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Calcidiol_metabolite_enzyme_sensor_GP'.\n",
      "Number of gene programs before filtering and combining: 1725.\n",
      "Number of gene programs after filtering and combining: 1575.\n",
      "Number of gene programs with probed genes: 489.\n",
      "Example gene programs: ['GREM1_ligand_targetgenes_GP', 'TSLP_ligand_targetgenes_GP', 'FGF7_ligand_targetgenes_GP', 'CDHR1_ligand_targetgenes_GP', 'INHA_ligand_targetgenes_GP'].\n",
      "Number of gene program target genes: 7476.\n",
      "Number of gene program source genes: 127.\n",
      "\n",
      "--- SPATIAL CONNECTIVITY ---\n",
      "Average number of edges per node: 8.0.\n",
      "Number of total edges: 77668.\n",
      "\n",
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, gene_expr_recon_dist: nb\n",
      "NODE LABEL METHOD -> one-hop-attention\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "GRAPH ENCODER -> n_input: 351, n_hidden: 489, n_latent: 489, n_addon_latent: 0, conv_layer: gcnconv, n_attention_heads: -, dropout_rate: 0.0\n",
      "DOT PRODUCT GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED GENE EXPRESSION DECODER -> n_input: 489, n_addon_input: 0, n_output: 702\n",
      "\n",
      "--- INITIALIZING TRAINER ---\n",
      "Number of training nodes: 17474\n",
      "Number of validation nodes: 1942\n",
      "Number of training edges: 73785\n",
      "Number of validation edges: 3883\n",
      "\n",
      "--- MODEL TRAINING ---\n",
      "Epoch 1/10 |ââ------------------| 10.0% val_auroc_score: 0.9352; val_auprc_score: 0.9040; val_best_acc_score: 0.8468; val_best_f1_score: 0.8670; train_kl_reg_loss: 0.1863; train_edge_recon_loss: 3.7677; train_gene_expr_recon_loss: 481.5321; train_group_lasso_reg_loss: 0.0000; train_global_loss: 485.4861; train_optim_loss: 485.4861; val_kl_reg_loss: 0.2422; val_edge_recon_loss: 2.4074; val_gene_expr_recon_loss: 477.8158; val_group_lasso_reg_loss: 0.0000; val_global_loss: 480.4653; val_optim_loss: 480.4653\n",
      "Epoch 2/10 |ââââ----------------| 20.0% val_auroc_score: 0.9400; val_auprc_score: 0.9110; val_best_acc_score: 0.8681; val_best_f1_score: 0.8827; train_kl_reg_loss: 0.2778; train_edge_recon_loss: 2.0691; train_gene_expr_recon_loss: 474.4266; train_group_lasso_reg_loss: 0.0000; train_global_loss: 476.7734; train_optim_loss: 476.7734; val_kl_reg_loss: 0.3263; val_edge_recon_loss: 1.9865; val_gene_expr_recon_loss: 477.3264; val_group_lasso_reg_loss: 0.0000; val_global_loss: 479.6393; val_optim_loss: 479.6393\n",
      "Epoch 3/10 |ââââââ--------------| 30.0% val_auroc_score: 0.9425; val_auprc_score: 0.9141; val_best_acc_score: 0.8966; val_best_f1_score: 0.9029; train_kl_reg_loss: 0.3104; train_edge_recon_loss: 1.7007; train_gene_expr_recon_loss: 475.5343; train_group_lasso_reg_loss: 0.0000; train_global_loss: 477.5454; train_optim_loss: 477.5454; val_kl_reg_loss: 0.2861; val_edge_recon_loss: 1.4838; val_gene_expr_recon_loss: 478.7305; val_group_lasso_reg_loss: 0.0000; val_global_loss: 480.5005; val_optim_loss: 480.5005\n",
      "Epoch 4/10 |ââââââââ------------| 40.0% val_auroc_score: 0.9487; val_auprc_score: 0.9304; val_best_acc_score: 0.8978; val_best_f1_score: 0.9046; train_kl_reg_loss: 0.3218; train_edge_recon_loss: 1.5093; train_gene_expr_recon_loss: 475.5772; train_group_lasso_reg_loss: 0.0000; train_global_loss: 477.4083; train_optim_loss: 477.4083; val_kl_reg_loss: 0.3204; val_edge_recon_loss: 1.3337; val_gene_expr_recon_loss: 477.5845; val_group_lasso_reg_loss: 0.0000; val_global_loss: 479.2387; val_optim_loss: 479.2387\n",
      "Epoch 5/10 |ââââââââââ----------| 50.0% val_auroc_score: 0.9375; val_auprc_score: 0.9150; val_best_acc_score: 0.8801; val_best_f1_score: 0.8890; train_kl_reg_loss: 0.3207; train_edge_recon_loss: 1.3968; train_gene_expr_recon_loss: 475.1459; train_group_lasso_reg_loss: 0.0000; train_global_loss: 476.8634; train_optim_loss: 476.8634; val_kl_reg_loss: 0.3185; val_edge_recon_loss: 1.4453; val_gene_expr_recon_loss: 477.7795; val_group_lasso_reg_loss: 0.0000; val_global_loss: 479.5433; val_optim_loss: 479.5433\n",
      "Epoch 6/10 |ââââââââââââ--------| 60.0% val_auroc_score: 0.9298; val_auprc_score: 0.9084; val_best_acc_score: 0.8636; val_best_f1_score: 0.8751; train_kl_reg_loss: 0.3161; train_edge_recon_loss: 1.3337; train_gene_expr_recon_loss: 475.5057; train_group_lasso_reg_loss: 0.0000; train_global_loss: 477.1555; train_optim_loss: 477.1555; val_kl_reg_loss: 0.3110; val_edge_recon_loss: 1.4631; val_gene_expr_recon_loss: 478.3358; val_group_lasso_reg_loss: 0.0000; val_global_loss: 480.1099; val_optim_loss: 480.1099\n",
      "Epoch 7/10 |ââââââââââââââ------| 70.0% val_auroc_score: 0.9310; val_auprc_score: 0.9046; val_best_acc_score: 0.8822; val_best_f1_score: 0.8921; train_kl_reg_loss: 0.3047; train_edge_recon_loss: 1.3345; train_gene_expr_recon_loss: 485.0889; train_group_lasso_reg_loss: 0.0000; train_global_loss: 486.7281; train_optim_loss: 486.7281; val_kl_reg_loss: 0.3084; val_edge_recon_loss: 1.3156; val_gene_expr_recon_loss: 479.0558; val_group_lasso_reg_loss: 0.0000; val_global_loss: 480.6797; val_optim_loss: 480.6797\n",
      "\n",
      "Reducing learning rate: metric has not improved more than 0.0 in the last 3 epochs.\n",
      "New learning rate is 0.001.\n",
      "\n",
      "Epoch 8/10 |ââââââââââââââââ----| 80.0% val_auroc_score: 0.9537; val_auprc_score: 0.9348; val_best_acc_score: 0.9057; val_best_f1_score: 0.9121; train_kl_reg_loss: 0.3253; train_edge_recon_loss: 1.1162; train_gene_expr_recon_loss: 474.5951; train_group_lasso_reg_loss: 0.0000; train_global_loss: 476.0365; train_optim_loss: 476.0365; val_kl_reg_loss: 0.3231; val_edge_recon_loss: 1.0579; val_gene_expr_recon_loss: 476.4102; val_group_lasso_reg_loss: 0.0000; val_global_loss: 477.7912; val_optim_loss: 477.7912\n",
      "Epoch 9/10 |ââââââââââââââââââ--| 90.0% val_auroc_score: 0.9490; val_auprc_score: 0.9299; val_best_acc_score: 0.9034; val_best_f1_score: 0.9095; train_kl_reg_loss: 0.3206; train_edge_recon_loss: 1.0897; train_gene_expr_recon_loss: 471.9264; train_group_lasso_reg_loss: 0.0000; train_global_loss: 473.3367; train_optim_loss: 473.3367; val_kl_reg_loss: 0.3067; val_edge_recon_loss: 1.0767; val_gene_expr_recon_loss: 475.5805; val_group_lasso_reg_loss: 0.0000; val_global_loss: 476.9638; val_optim_loss: 476.9638\n",
      "Epoch 10/10 |ââââââââââââââââââââ| 100.0% val_auroc_score: 0.9519; val_auprc_score: 0.9345; val_best_acc_score: 0.9039; val_best_f1_score: 0.9085; train_kl_reg_loss: 0.3002; train_edge_recon_loss: 1.0649; train_gene_expr_recon_loss: 472.7284; train_group_lasso_reg_loss: 0.0000; train_global_loss: 474.0935; train_optim_loss: 474.0935; val_kl_reg_loss: 0.2982; val_edge_recon_loss: 1.0505; val_gene_expr_recon_loss: 475.3780; val_group_lasso_reg_loss: 0.0000; val_global_loss: 476.7266; val_optim_loss: 476.7266\n",
      "\n",
      "Model training finished after 5 min 50 sec.\n",
      "Using best model state, which was in epoch 10.\n",
      "\n",
      "--- MODEL EVALUATION ---\n",
      "Val AUROC score: 0.9527\n",
      "Val AUPRC score: 0.9381\n",
      "Val best accuracy score: 0.9041\n",
      "Val best F1 score: 0.9087\n",
      "Val MSE score: 0.8902\n",
      "--------------------\n",
      "\n",
      "--------------------\n",
      "\n",
      "--- SPATIAL CONNECTIVITY ---\n",
      "Average number of edges per node: 8.0.\n",
      "Number of total edges: 77668.\n",
      "\n",
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, gene_expr_recon_dist: nb\n",
      "NODE LABEL METHOD -> one-hop-attention\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "GRAPH ENCODER -> n_input: 351, n_hidden: 489, n_latent: 489, n_addon_latent: 0, conv_layer: gcnconv, n_attention_heads: -, dropout_rate: 0.0\n",
      "DOT PRODUCT GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED GENE EXPRESSION DECODER -> n_input: 489, n_addon_input: 0, n_output: 702\n",
      "\n",
      "--- INITIALIZING TRAINER ---\n",
      "Number of training nodes: 17474\n",
      "Number of validation nodes: 1942\n",
      "Number of training edges: 73785\n",
      "Number of validation edges: 3883\n",
      "\n",
      "--- MODEL TRAINING ---\n",
      "Epoch 1/10 |ââ------------------| 10.0% val_auroc_score: 0.9392; val_auprc_score: 0.9014; val_best_acc_score: 0.8237; val_best_f1_score: 0.8501; train_kl_reg_loss: 0.1911; train_edge_recon_loss: 3.7138; train_gene_expr_recon_loss: 480.3823; train_group_lasso_reg_loss: 0.0000; train_global_loss: 484.2872; train_optim_loss: 484.2872; val_kl_reg_loss: 0.2937; val_edge_recon_loss: 2.8251; val_gene_expr_recon_loss: 476.8483; val_group_lasso_reg_loss: 0.0000; val_global_loss: 479.9671; val_optim_loss: 479.9671\n",
      "Epoch 2/10 |ââââ----------------| 20.0% val_auroc_score: 0.9574; val_auprc_score: 0.9309; val_best_acc_score: 0.8809; val_best_f1_score: 0.8934; train_kl_reg_loss: 0.2880; train_edge_recon_loss: 2.1748; train_gene_expr_recon_loss: 473.1313; train_group_lasso_reg_loss: 0.0000; train_global_loss: 475.5942; train_optim_loss: 475.5942; val_kl_reg_loss: 0.3266; val_edge_recon_loss: 2.0109; val_gene_expr_recon_loss: 475.9832; val_group_lasso_reg_loss: 0.0000; val_global_loss: 478.3207; val_optim_loss: 478.3207\n",
      "Epoch 3/10 |ââââââ--------------| 30.0% val_auroc_score: 0.9627; val_auprc_score: 0.9433; val_best_acc_score: 0.8963; val_best_f1_score: 0.9058; train_kl_reg_loss: 0.3200; train_edge_recon_loss: 1.7810; train_gene_expr_recon_loss: 473.7974; train_group_lasso_reg_loss: 0.0000; train_global_loss: 475.8984; train_optim_loss: 475.8984; val_kl_reg_loss: 0.3275; val_edge_recon_loss: 1.6998; val_gene_expr_recon_loss: 476.4984; val_group_lasso_reg_loss: 0.0000; val_global_loss: 478.5257; val_optim_loss: 478.5257\n"
     ]
    }
   ],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.05102] # [0.03875][0.05102]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"] #\"nb\"\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-attention\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"] # [\"gatv2conv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [0] # 1,2,4,8, 16\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [5]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_encoder_decoder_arch2\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=2,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8686482-3dfc-45b8-a716-66a9cde9dbf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"radius\"] = [0.03875]\n",
    "hyperparam_option_dict[\"include_edge_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"include_gene_expr_recon_loss\"] = [True]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"zinb\"] #\"nb\"\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-norm\", \"one-hop-sum\", \"one-hop-attention\"]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\", \"gatv2conv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1]\n",
    "hyperparam_option_dict[\"n_addon_gps\"] = [0]\n",
    "hyperparam_option_dict[\"n_epochs\"] = [10]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [5]\n",
    "hyperparam_option_dict[\"lr\"] = [0.01]\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "hyperparam_option_dict[\"lambda_l1_addon\"] = [0]\n",
    "\n",
    "experiment_name = \"benchmark_encoder_decoder_arch\"\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                            hyperparam_option_dict=hyperparam_option_dict,\n",
    "                            n_iters=3,\n",
    "                            experiment_name=experiment_name,\n",
    "                            gp_mask=\"combined_priors\",\n",
    "                            save_model=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de367f1c-790c-47de-a17f-3be73c7c86f4",
   "metadata": {},
   "source": [
    "### 2.4. Benchmark Loss Inclusions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7bbe6f5-9122-428a-9595-be5bdd502546",
   "metadata": {},
   "source": [
    "### 2.5. Benchmark De-Novo GPs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
